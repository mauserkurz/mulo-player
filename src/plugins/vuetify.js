import Vue from 'vue';
import Vuetify from 'vuetify';
import { THEME as theme } from '@/constants';

Vue.use(Vuetify);

export default new Vuetify({ theme });
