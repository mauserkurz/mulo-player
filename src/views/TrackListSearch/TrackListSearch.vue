<template>
  <v-menu
    ref="menu"
    v-model="isDatePickerShown"
    :close-on-content-click="false"
    :return-value.sync="value"
    :disabled="!isDatePickerAvailable"
    class="track-list-search"
    transition="scale-transition"
    offset-y
    max-width="290px"
    min-width="290px">
    <template v-slot:activator="{ on }">
      <v-text-field
        v-on="on"
        :value="value"
        :readonly="isDatePickerAvailable"
        prepend-inner-icon="search"
        flat
        solo-inverted
        clearable
        label="Search"
        @input="onSearchInput"
        @click:clear="onSearchInput('')"/>
    </template>

    <v-date-picker
      :value="value"
      :max="today"
      :show-current="false"
      no-title
      @input="onSearchInput">
      <v-spacer/>

      <v-btn
        color="primary"
        text
        @click="switchDatePicker(false)">
        Cancel
      </v-btn>

      <v-btn
        color="primary"
        text
        @click="$refs.menu.save(value)">
        OK
      </v-btn>
    </v-date-picker>
  </v-menu>
</template>

<script>
import { FILTER_TRACK_TYPE_KEYS } from '@/constants';

const DATE_FILTER_TYPE_LIST = [
  FILTER_TRACK_TYPE_KEYS.NEW,
  FILTER_TRACK_TYPE_KEYS.OLD,
  FILTER_TRACK_TYPE_KEYS.DURING,
];

export default {
  name: 'TrackListSearch',

  props: {
    value: {
      type: String,
      required: true,
    },

    filterType: {
      type: String,
      required: true,
    },
  },

  data() {
    return { isDatePickerShown: false };
  },

  computed: {
    isDatePickerAvailable() {
      return DATE_FILTER_TYPE_LIST.includes(this.filterType);
    },

    today() {
      return new Date().toISOString();
    },
  },

  watch: {
    filterType() {
      const state = this.isDatePickerAvailable;

      if (!state) {
        this.onSearchInput('');
      }
      this.switchDatePicker(state);
    },
  },

  methods: {
    switchDatePicker(state) {
      this.isDatePickerShown = state;
    },

    onSearchInput(value) {
      this.$emit('input', value || '');
    },
  },
};
</script>

<style scoped lang="less">
.track-list-search {}
</style>
