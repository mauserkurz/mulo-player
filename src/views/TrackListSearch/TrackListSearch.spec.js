import { mount, shallowMount, createLocalVue } from '@vue/test-utils';
import vuetify from '@/plugins/vuetify';
import { FILTER_TRACK_TYPE_KEYS } from '@/constants';
import TrackListSearch from './TrackListSearch.vue';

const localVue = createLocalVue();

describe('Component TrackListSearch', () => {
  const defaultProps = {
    value: '',
    filterType: FILTER_TRACK_TYPE_KEYS.TEXT,
  };
  const createWrapper = ({ isWithoutStubs = false, propsData = {}, ...options } = {}) => {
    const renderer = isWithoutStubs ? mount : shallowMount;
    const wrapper = renderer(TrackListSearch, {
      localVue,
      vuetify,
      propsData: Object.assign({}, defaultProps, propsData),
      ...options,
    });

    return { wrapper };
  };

  describe('snapshots', () => {
    it('text search should match snapshot', () => {
      const { wrapper } = createWrapper({ isWithoutStubs: true });

      expect(wrapper.html()).toMatchSnapshot();
    });

    it('date search with opened date picker should match snapshot', async () => {
      const { wrapper } = createWrapper({ isWithoutStubs: true });

      wrapper.setProps({ filterType: FILTER_TRACK_TYPE_KEYS.NEW });
      await wrapper.vm.$nextTick();
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('events', () => {
    it('should emit text input value', async () => {
      const { wrapper } = createWrapper({ isWithoutStubs: true });
      const input = wrapper.find('input[type="text"]');
      const value = 'Band 1';

      input.element.value = value;
      input.trigger('input');
      await wrapper.vm.$nextTick();

      expect(wrapper.emitted().input).toEqual([[value]]);
    });

    it('after change filterType to text should hide date picker', async () => {
      const { wrapper } = createWrapper({
        isWithoutStubs: true,
        propsData: { filterType: FILTER_TRACK_TYPE_KEYS.DURING },
      });

      wrapper.setProps({ filterType: FILTER_TRACK_TYPE_KEYS.TEXT });
      await wrapper.vm.$nextTick();
      expect(wrapper.find('.v-picker').exists()).toBe(false);
    });

    it('should emit input with "" after click on clear icon', async () => {
      const { wrapper } = createWrapper({ isWithoutStubs: true, propsData: { value: 'Band 0' } });

      wrapper.find('.v-text-field').vm.$emit('click:clear');
      await wrapper.vm.$nextTick();
      expect(wrapper.emitted().input).toEqual([['']]);
    });

    it('after change filterType to date type should show date picker', async () => {
      const { wrapper } = createWrapper({
        isWithoutStubs: true,
        propsData: { filterType: FILTER_TRACK_TYPE_KEYS.TEXT },
      });

      wrapper.setProps({ filterType: FILTER_TRACK_TYPE_KEYS.DURING });
      await wrapper.vm.$nextTick();
      expect(wrapper.find('.v-picker').exists()).toBe(true);
    });

    it('on select date should emit date picker value', async () => {
      const { wrapper } = createWrapper({
        isWithoutStubs: true,
        propsData: { filterType: FILTER_TRACK_TYPE_KEYS.TEXT },
      });
      const value = '2020-04-15';

      wrapper.setProps({ filterType: FILTER_TRACK_TYPE_KEYS.DURING });
      await wrapper.vm.$nextTick();
      wrapper.find('.v-picker').vm.$emit('input', value);
      expect(wrapper.emitted().input).toEqual([[''], [value]]);
    });

    it('after click on Cancel button should hide date picker', async () => {
      const { wrapper } = createWrapper({
        isWithoutStubs: true,
        propsData: { filterType: FILTER_TRACK_TYPE_KEYS.TEXT },
      });

      wrapper.setProps({ filterType: FILTER_TRACK_TYPE_KEYS.DURING });
      await wrapper.vm.$nextTick();
      wrapper.findAll('.v-btn').at(33).trigger('click');
      await wrapper.vm.$nextTick();

      expect(wrapper.find('.v-menu__content').element.style.display)
        .toBe('none');
    });

    it('after click on Ok button should hide date picker', async () => {
      const { wrapper } = createWrapper({
        isWithoutStubs: true,
        propsData: { filterType: FILTER_TRACK_TYPE_KEYS.TEXT },
      });

      wrapper.setProps({ filterType: FILTER_TRACK_TYPE_KEYS.DURING });
      await wrapper.vm.$nextTick();
      wrapper.findAll('.v-btn').at(34).trigger('click');
      await wrapper.vm.$nextTick();

      expect(wrapper.find('.v-menu__content').element.style.display)
        .toBe('none');
    });
  });
});
