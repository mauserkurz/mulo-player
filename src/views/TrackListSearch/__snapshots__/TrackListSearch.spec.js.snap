// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component TrackListSearch snapshots date search with opened date picker should match snapshot 1`] = `
<div class="v-menu track-list-search">
  <div class="v-input v-input--is-readonly theme--light v-text-field v-text-field--single-line v-text-field--solo v-text-field--solo-inverted v-text-field--solo-flat v-text-field--enclosed">
    <div class="v-input__control">
      <div class="v-input__slot">
        <div class="v-input__prepend-inner">
          <div class="v-input__icon v-input__icon--prepend-inner"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light">search</i></div>
        </div>
        <div class="v-text-field__slot"><label for="input-15" class="v-label theme--light" style="left: 0px; position: absolute;">Search</label><input id="input-15" type="text" readonly="readonly"></div>
        <div class="v-input__append-inner">
          <div class="v-input__icon v-input__icon--"><i role="button" tabindex="0" class="v-icon notranslate v-icon--link material-icons theme--light"></i></div>
        </div>
      </div>
      <div class="v-text-field__details">
        <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
      </div>
    </div>
  </div>
  <div role="menu" class="v-menu__content theme--light menuable__content__active " style="max-height: auto; min-width: 290px; max-width: 290px; top: 12px; left: 0px; transform-origin: top left; z-index: 8; display: none; z-index: 8;">
    <div class="v-picker v-card v-picker--date theme--light">
      <div class="v-picker__body v-picker__body--no-title theme--light" style="width: 290px;">
        <div>
          <div class="v-date-picker-header theme--light"><button type="button" class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-left theme--light"></i></span></button>
            <div class="v-date-picker-header__value">
              <div class="accent--text"><button type="button">May 2020</button></div>
            </div><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-right theme--light"></i></span></button>
          </div>
          <div class="v-date-picker-table v-date-picker-table--date theme--light">
            <table>
              <thead>
                <tr>
                  <th>S</th>
                  <th>M</th>
                  <th>T</th>
                  <th>W</th>
                  <th>T</th>
                  <th>F</th>
                  <th>S</th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  <td></td>
                  <td></td>
                  <td></td>
                  <td></td>
                  <td></td>
                  <td><button type="button" class="v-btn v-btn--text v-btn--rounded theme--light">
                      <div class="v-btn__content">1</div>
                    </button></td>
                  <td><button type="button" class="v-btn v-btn--text v-btn--rounded theme--light">
                      <div class="v-btn__content">2</div>
                    </button></td>
                </tr>
                <tr>
                  <td><button type="button" class="v-btn v-btn--text v-btn--rounded theme--light">
                      <div class="v-btn__content">3</div>
                    </button></td>
                  <td><button type="button" class="v-btn v-btn--text v-btn--rounded theme--light">
                      <div class="v-btn__content">4</div>
                    </button></td>
                  <td><button type="button" class="v-btn v-btn--text v-btn--rounded theme--light">
                      <div class="v-btn__content">5</div>
                    </button></td>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">6</div>
                    </button></td>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">7</div>
                    </button></td>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">8</div>
                    </button></td>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">9</div>
                    </button></td>
                </tr>
                <tr>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">10</div>
                    </button></td>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">11</div>
                    </button></td>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">12</div>
                    </button></td>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">13</div>
                    </button></td>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">14</div>
                    </button></td>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">15</div>
                    </button></td>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">16</div>
                    </button></td>
                </tr>
                <tr>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">17</div>
                    </button></td>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">18</div>
                    </button></td>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">19</div>
                    </button></td>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">20</div>
                    </button></td>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">21</div>
                    </button></td>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">22</div>
                    </button></td>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">23</div>
                    </button></td>
                </tr>
                <tr>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">24</div>
                    </button></td>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">25</div>
                    </button></td>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">26</div>
                    </button></td>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">27</div>
                    </button></td>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">28</div>
                    </button></td>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">29</div>
                    </button></td>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">30</div>
                    </button></td>
                </tr>
                <tr>
                  <td><button type="button" class="v-btn v-btn--flat v-btn--text v-btn--rounded v-btn--disabled theme--light" disabled="">
                      <div class="v-btn__content">31</div>
                    </button></td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
      <div class="v-picker__actions v-card__actions v-picker__actions--no-title">
        <div class="spacer"></div> <button type="button" class="v-btn v-btn--flat v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content">
      Cancel
    </span></button> <button type="button" class="v-btn v-btn--flat v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content">
      OK
    </span></button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Component TrackListSearch snapshots text search should match snapshot 1`] = `
<div class="v-menu track-list-search">
  <div class="v-input theme--light v-text-field v-text-field--single-line v-text-field--solo v-text-field--solo-inverted v-text-field--solo-flat v-text-field--enclosed">
    <div class="v-input__control">
      <div class="v-input__slot">
        <div class="v-input__prepend-inner">
          <div class="v-input__icon v-input__icon--prepend-inner"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light">search</i></div>
        </div>
        <div class="v-text-field__slot"><label for="input-4" class="v-label theme--light" style="left: 0px; position: absolute;">Search</label><input id="input-4" type="text"></div>
        <div class="v-input__append-inner">
          <div class="v-input__icon v-input__icon--"><i role="button" tabindex="0" class="v-icon notranslate v-icon--link material-icons theme--light"></i></div>
        </div>
      </div>
      <div class="v-text-field__details">
        <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
      </div>
    </div>
  </div>
  <div role="menu" class="v-menu__content theme--light " style="max-height: auto; min-width: 290px; max-width: 290px; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"></div>
</div>
`;
